### DISCUSSION TITLE
Feature request: be able to add limit on a block type in a Matrix field

### DISCUSSION THREAD

**Discussion by bzin** [craftcms/cms#8596] [Ideas]
### Description

So I am using `Matrix` field to create data structure for twig components that stack on top of each other in a generic page. I do not want to create any limitation on the number of blocks but it would be  interesting and useful to be able to set a limit also on a specific block type. 

Would this be something you see being added in a future release?

### Additional info

- Craft version: Craft Pro 3.0.26.1
- PHP version: 7.2.7

**Comment by brandonkelly** [comment]
Technically this is a duplicate of #826 but I’ll close that one instead as it never received any new activity since we switched from UserVoice to GH issues.

It’s easy enough to add it, but I’d like to better understand the use case. What sort of block within a Matrix-powered page builder would need to be limited to only use X times?

**Comment by richardcalahan** [comment]
Often times I'll use a matrix field to build a modular page. Each matrix block is a module, and the module types would include: hero, image + copy, video, map, etc. 

"Hero" or "Map" might only make sense to be included in the page once, but still orderable within all modules.

**Comment by bzin** [comment]
My usage case is much like @richardcalahan is saying, with the difference that I also use Matrix to use custom (twig) components, let's say you have `Header`, `Banner`, `Social Feed`, `Text + Media`, etc...

Using this blocks/components, I might want to limit the usage of for the `Banner` and `Header` components, but I do not wish to have the limitation on creating the ones that are more story and content driven.

**Comment by Anubarak** [comment]
Another use case are social media channels. Imagine you have a channel of persons and each person should be able to insert social media links or whatever information we require (different api tokens and what not)

Of course one could argue this is better off as a separate field type but I like to use a matrix where the different block types represents different platforms and limit them via JavaScript to one since it doesn't make sense to show the Facebook icon twice next to each other in a teaser/block/header..

Matrix blocks are not always used as a page builder. It's their main purpose of course but you can use them as a simple collection container as well

**Comment by bzin** [comment]
That reminds another usage I give to this field. Depending on the project I might use a Matrix on a Global set to create a `Menu` and `Footer` elements. Moreover, in this Matrix, I might want to limit certain buttons, links, text, etc...

**Comment by piotrpog** [comment]
Just recently i also had situation when this would be usefull - dont need multiple sections "sign up for newsletter", but i need multiple sections "products from specific  category" - for example.

**Comment by IndigoViking** [comment]
I've needed this on several occasions. We use matrix blocks for main content or homepage blocks and typically only want one of each block, or sometimes just one block of a few types.

**Comment by BridgingUnit** [comment]
Yep, could really do with this.

Some use cases I've hit recently:

Homepage with matrix with blocks (because client would like to reorder output of sections), but each block should only be displayed once.

Matrix used for Slideshow, or display of any set of same blocks (e.g. Bios etc), but where there's a block used for headline/intro first. That should only display once, but block type of slide (or bio etc) should be multiple.

Matrix used to display set of 4 panels within a page, but should only use each once. Reason being each has slightly different fields to provide content. Again, using Matrix so client can rearrange the order of them.

**Comment by narration-sd** [comment]
Seems to want to occur to me that if you do this, it would be good to have some flexibility in what accounts for uniqueness. 

Type is one factor, but a string or numeric match (make use  of PHP''s looseness??) on a chosen field could be another. Thus you could assure you didn't try to make an employee bio, a product card, etc. over again, due to differing or updated information.

Likely a field opens here also of providing a facility to _find_ duplicated blocks, remind you if you try to save them, etc....

**Comment by bzin** [comment]
Any idea if this will see the light of day @brandonkelly? Missing this so much! 🙏

**Comment by daltonrooney** [comment]
This can be done with [MatrixMate](https://github.com/vaersaagod/matrixmate). Invaluable plugin, IMHO. Look for the 'maxLimit' config option for each block type.

**Comment by siffring** [comment]
I could use this on a specification table for a product. Imagine a series of key/value pairs that describe specs for a product (e.g. height: 10, weight: 20, etc.).

The layout we're working with has a one column option for products with a few specs and a two column layout for products with a lot of specs.

With a block type limit I could have a block type of "column break" that can only be used once. Without the limit I'll end up creating two matrix fields for "Specs first column" and "Specs second column" which is a little less elegant.

**Comment by terryupton** [comment]
Any plans to bring this to Craft 4? @brandonkelly

**Reply by BleonaHalimi** [reply]
I believe this feature would greatly benefit users who frequently work with Matrix fields.

**Comment by npruzaniec** [comment]
With the sunsetting of Matrixmate for Craft 5 - is there currently a way to do this natively or another solution anyone has found? Losing this capability is currently the only thing keeping us from upgrading from 4 - 5.

**Comment by anita-chouhan** [comment]
We also use this feature a lot where matrix fields are used for the main content, and sometimes a block type is needed only once or twice for a page, like the _banner_ or the _map_ block. 

Any plans to bring this to Craft 5? @brandonkelly

**Reply by arentsen** [reply]
Yeah, I wrote this in another thread, but copy it here. It is such a useful feature.


_It would be great if it would be possible to restrict how many times an entry can be added in a matrix field. Now, only the total amount of entries can be restricted, but not how many times an entry can be used._

_This is great if a page only contains certain modules once, but giving the user the flexibility to add/disable items and change their order. This restriction should then be added within the matrix field settings and not in the entry settings. In the old Matrix setup in Craft 4 it was possible to say how many times a block could be used._

**Reply by mmikkel** [reply]
@arentsen FTR, this has never been a core feature – you were likely using MatrixMate or Spoon for Craft 4?

**Comment by daltonrooney** [comment]
This can be accomplished pretty easily with a custom site module, see https://nystudio107.com/blog/custom-matrix-block-validation-rules

**Reply by mmikkel** [reply]
Custom validation rules can be used to prevent authors from saving entries with a disallowed number of nested entries using a specific entry type. Which is an OK solution, but pretty frustrating AX; ideally, the UI would prevent authors from creating those entries in the first place.

**Comment by furioursus** [comment]
Hooboy, i’m really running into a need for it myself. My choices are currently Neo or help explain to a governmental body that there are specific blocks they should only include once.

For the meantime i will be limiting the number of times a specific entry type is rendered on the frontend, but it would be great to ensure the content admin doesn’t accidentally (or intentionally) add multiple maps or banners in our content builders.

**Comment by piotrpog** [comment]
@furioursus I guess you could add custom validation rules to your entry using matrix field, by using module or plugin.

**Reply by furioursus** [reply]
That requires additional maintenance and control, which are fine for a short term solution but do not necessarily work as something I would want to leave the client with after i exit. The entire point of the request is to point out why we all would need it.
